### ISSUE TITLE
JDK's ForkJoinTask#inForkJoinPool() returns false when inside akka's ForkJoinPool

### ISSUE THREAD

**Issue by mkurz** [akka/akka#23662]
or: "Java parallel Streams won't use akka's fork-join pools"
---
I am using the Play Framework (Java - not Scala, latest version 2.6.5).

Turns out akka is using it's own implementation of [ForkJoinPool](https://github.com/akka/akka/blob/v2.5.4/akka-actor/src/main/java/akka/dispatch/forkjoin/ForkJoinPool.java), [ForkJoinTask](https://github.com/akka/akka/blob/v2.5.4/akka-actor/src/main/java/akka/dispatch/forkjoin/ForkJoinTask.java) and [ForkJoinWorkerThread](https://github.com/akka/akka/blob/v2.5.4/akka-actor/src/main/java/akka/dispatch/forkjoin/ForkJoinWorkerThread.java) - and not using the ones that come with JDK-8 (unlike Scala, which [switched](https://github.com/scala/bug/issues/9381) to the JDK ones [in 2.12](https://github.com/scala/scala/pull/4629)). For various reasons, one of them is performance, as already discussed in #18262, #22560, #21910 and #21101.

As you probably know Play uses a fork-join executor - an akka dispatcher - to handle all the requests,
 see its [reference-overrides.conf](https://github.com/playframework/playframework/blob/2.6.5/framework/src/play/src/main/resources/play/reference-overrides.conf#L21-L35). Therefore a Play action method is executed within such an akka fork-join pool.

So far so good.

The point is when I use a Java **parallel** Stream inside a Play action method I would expect that all the tasks of such a parallel stream (`map`, `filter`, etc.) would **not** run within `ForkJoinPool.commonPool()` but within the current akka fork-join pool because I am already "inside" *that* play fork-join pool. Look at the [javadoc](http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinTask.html#fork--) of JDK's `ForkJoinTask#fork()`
> Arranges to asynchronously execute this task in the pool the current task is running in, if applicable, or using the ForkJoinPool.commonPool() if not inForkJoinPool()

So `fork` checks if the current caller thread is in a ForkJoinPool. And I would expect "yes, I am `in`a`ForkJoinPool`". And this is how JDK `inForkJoinPool` is implemented:
```java
    public static boolean inForkJoinPool() {
        return Thread.currentThread() instanceof ForkJoinWorkerThread;
    }
```
So the JDK checks if the current caller thread is an instance of `java.util.concurrent.ForkJoinWorkerThread` and if that's the case it uses the thread's fork-join pool to execute the Java Stream tasks. But because akka uses it's own `ForkJoinWorkerThread` which **is *not* an instance of JDK's `ForkJoinWorkerThread`** the method returns `false` and therefore we end up executing our parallel stream tasks in the `ForkJoinPool.commonPool()`.
And for me that seems wrong, because, again, we are `in`a`ForkJoinPool`!

I hope you get the problem here: I would expect that inside a Play action the tasks of a Java parallel stream would be executed inside the current (akka) fork-join pool and not the `ForkJoinPool.commonPool()`. That is what I am expecting when tuning Play's default dispatcher.
(Right now I would have to wrap all the parallel stream inside an own JDK ForkJoinPool to take control over how they execute, like described in [this stackoverflow answer](https://stackoverflow.com/a/22269778/810109)).

**Comment by ktoso** [comment]
Hi there, I'm not sure what outcome of this ticket you are expecting Matthias. Yes, this is how the JDK's pool works. You could configure Akka and Play to use the JDK provided FJP, though then, as you've found already in the linked tickets, we'll take a not insignificant performance hit.

I think there's some options coming in JDK9 (to be released tomorrow), which may allow us to use that pool as default instead, but we have not benchmarked it yet.

**Comment by jrudolph** [comment]
@mkurz I understand that it could be convenient if parallel streams could use Akka's executor. On the other hand, Akka's usage of the fork join pool is more of an implementation detail than a promise to provide services to third-party APIs (like the parallel streams API in this case).  It seems quite brittle to rely on these kinds of certain global constellations to align for your application to work optimally.

Have you filed an issue with the JDK? I think in this case the parallel streams API is to blame when it doesn't provide a proper API to select a context for execution but relies on global state and internals to choose where to run.

**Comment by mkurz** [comment]
@ktoso I opened this issue because for me - at the first glance - it doesn't really seem logical that a Java parallel Stream's tasks are not executed within the Play/akka FJP *even though (in my opinion) the requirement to do so is actually fulfilled*: **Are we in a FJP? Yes we are**. So, without looking at its implementation, I would expect JDK's `inForkJoinPool` to return `true`, even when the FJP is an akka one.
Do you understand my point of view? I don't think my expectation is completely unrealistic. 

So, based on my FJP expectation, the outcome of this ticket I am expecting is that JDK's `inForkJoinPool` will return `true` when I am inside an akka FJP.
So what can be done on akka's side to meet my expectation? What comes in my mind is to make the akka FJP a sub-class of JDK's FJP, or, like you said, just use the JDK provided FJP inside akka - which, right now, isn't preferable because of the performance degradation this would result in.

So the question for me is: Is my expectation regarding FJP justified?

**Comment by mkurz** [comment]
@jrudolph OK, I agree with you. Relying on the implementation details of `ForkJoinTask#fork()` / `ForkJoinTask#inForkJoinPool()` isn't probably the best idea and JDK should provide a way of explicitly setting the context for execution the parallel Streams' tasks. I will check if there are already any open JDK bugs related to this later today.

Anyway, I also think akka should make use of JDK 9 FJP **if** performance is acceptable. Looking forward to those benchmarks.

**Comment by mkurz** [comment]
However... leaving the Java Streams API beside... it still buggers me a bit that `java.util.concurrent.ForkJoinTask.inForkJoinPool()` returns `false` *even when a thread is in a FJP* - even though it's a akka one.
I think *that* is the whole point I want to make here.

**Comment by ktoso** [comment]
>  tasks are not executed within the Play/akka FJP even though (in my opinion) the requirement to do so is actually fulfilled: Are we in a FJP? Yes we are. So, without looking at its implementation, I would expect JDK's inForkJoinPool to return true, even when the FJP is an akka one.

Please note that's your interpretation, and not the actual requirement. The actual requirement is written down in code and is: "Are we running in JDK's FJP?" So the answer is no :-)

Now we may of course argue if that's the right check to do there, with no option for other pools to "opt in", however that would be discussed on the JDK's issue tracker, as it's a decision made by the JDK team and in that codebase. If the issue has not been raised yet there, I suggest you do so, explaining the use-cases that this would support.

Johannes raises another good point that this perhaps is not what you'd want in an Akka-under-the-hood-powered application to begin with.

You can also just change the dispatcher used to be , take the performance hit, and have it work if you wanted to.

Making the Akka FJP extend the Java one is not really something we'd want to pursue, I tried this at some point ant the maintenance burden and complexity does not pay for itself to do such a weird trick.

Summing up: we agree with each other it seems – the behaviour is what it is. It's defined in the JDK, we can't change it. We don't want to use the default pool, at least not on JDK8. Sharing the Akka pool with Java's streams may not be what you want in any case. I'm not sure there's an action to be taken right now; With JDK9 coming out *tomorrow*, we should rather look forward and investigate that Pool's performance for our use case (AFAIR it exposes a flat about spinning that we needed).

Having that said, we'll revisit this topic naturally sooner or later and I'd like to propose closing this issue, as there is not immediate action to resolve it - at least not in the form as has been stated here.

**Comment by mkurz** [comment]
@ktoso and @jrudolph: Thanks for taking the time writing down your thoughts on this topic. I agree that we agree with each other :smile: - it is how it is right now.

Side note: The funny thing is if akka switches to JDK's (post 8) FJP someday, then 

> Sharing the Akka pool with Java's streams may not be what you want in any case.

will become the default behaviour inside Play action methods. (However I think to be aware of this behaviour is an app-developer's responsibility anyway - he/she should understand in which context a task/thread will run).
Nevertheless I think removing akka's custom FJP and switching to the JDK 9 FJP - if performance goals are meet - is the way to go in long term.

**Comment by mkurz** [comment]
(Not sure if I should close or leave issue open as a reminder to revisit JDK-9 FJP)

**Comment by ktoso** [comment]
Closing is fine, thanks - it's something we are keeping in the backs of our heads :)

**Comment by mkurz** [comment]
Alright then, closing :smile:

**Comment by iksaif** [comment]
should be fixed by https://github.com/akka/akka/commit/c90f5720b3ec3508e1d07fd6089217a20ef25c75#diff-410abf8aac6a89a01b05db2cbcf1765c

**Comment by mkurz** [comment]
@iksaif Thanks, I am already aware that #26893 got merged but thanks for the pointer anyway!
